src/django_components/component.py

```diff
@@ -630,17 +628,21 @@ def _render_impl(
         type: RenderType = "document",
         render_dependencies: bool = True,
     ) -> str:
+        # NOTE: We must run validation before we normalize the slots, because the normalization
+        #       wraps them in functions.
+        self._validate_inputs(args or (), kwargs or {}, slots or {})
+
         # Allow to provide no args/kwargs/slots/context
         args = cast(ArgsType, args or ())
         kwargs = cast(KwargsType, kwargs or {})
-        slots = cast(SlotsType, slots or {})
+        slots_untyped = self._normalize_slot_fills(slots or {}, escape_slots_content)
```

Comment:
The slot fills are normalize in two steps:
1. They are normalized to functions. So if user supplies a plain string via calling 
`Component.render(slots={...})`, then we convert it to dummy function that returns the said function.
2. Slot fill functions are then wrapped in a `Slot` instance.

In this MR I added the step 2.

The `Slot` class behaves as a function on the surface (calling the associated slot fill function), so interface doesn't change. But it allows us to associate extra metadata with those slot fill functions.

Moreover, `Slot` class is used to make it possible to pass the slot fill functions as variables within the Django template.
   - The `Slot` class defines a `do_not_call_in_templates = True` property, which tells Django to NOT automatically call the function

See the definition here
https://github.com/EmilStenstrom/django-components/pull/758/files#diff-c7f4341cbd58fb82a995fdba37703267d2dbbdea83ac86c7cb2a67c258fd835cR67